File: java/src/main/java/com/purescraps/purehtml/transformers/types/JoinTransformer.java
package com.purescraps.purehtml.transformers.types;

import java.util.List;
import java.util.stream.Collectors;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

/**
 * join: Join a list of values into a string
 *
 * Usage:
 * join(): joins with an empty string
 * join(delimiter): joins elements using `delimiter`
 */
public class JoinTransformer extends Transformer {
    private List<String> args;

    public JoinTransformer(List<String> args) {
        this.args = args;
    }

    public JoinTransformer() {
        this.args = null;
    }

    public void setArgs(List<String> args) {
        this.args = args;
    }

    public String getName() {
        return "join";
    }

    @Override
    public Object transform(TransformParams params) {
        Object val = params.getVal();

        if (val == null) {
            return null;
        }

        if (!(val instanceof List)) {
            throw new IllegalArgumentException("join: expected array. Got " + val);
        }

        String delimiter = (args != null && !args.isEmpty()) ? args.get(0) : "";

        return ((List<?>) val).stream()
                .map(String::valueOf)
                .collect(Collectors.joining(delimiter));
    }
}

File: python/purehtml/transformers/types/join_transformer.py
from abc import ABC
from typing import List

from purehtml.transformers.transform_params import TransformParams
from purehtml.transformers.transformer import Transformer


class JoinTransformer(Transformer, ABC):
    """
    join: Join a list of values into a string

    Usage:
    join(): joins with an empty string
    join(delimiter): joins elements using `delimiter`
    """

    def __init__(self, args: List[str] = None):
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        self.args = args

    def get_name(self) -> str:
        return "join"

    def transform(self, params: TransformParams):
        val = params.get_val()

        if val is None:
            return None

        if not isinstance(val, (list, tuple)):
            raise TypeError(f"join: expected array. Got {val!r}")

        delimiter = self.args[0] if self.args else ""

        return delimiter.join(str(item) for item in val)
